airbyte-integrations/connectors/source-mailchimp/source_mailchimp/source.py

```diff
@@ -21,67 +21,201 @@
-
-from .client import Client
-
-
-class SourceMailchimp(Source):
-    """
-    Mailchimp API Reference: https://mailchimp.com/developer/api/
-    """
-
-    def check(self, logger: AirbyteLogger, config: json) -> AirbyteConnectionStatus:
-        client = self._client(config)
-        alive, error = client.health_check()
-        if not alive:
-            return AirbyteConnectionStatus(status=Status.FAILED, message=f"{error.title}: {error.detail}")
-
-        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
-
-    def discover(self, logger: AirbyteLogger, config: json) -> AirbyteCatalog:
-        client = self._client(config)
-
-        return AirbyteCatalog(streams=client.get_streams())
-
-    def read(
-        self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state: Dict[str, any]
-    ) -> Generator[AirbyteMessage, None, None]:
-        client = self._client(config)
-
-        logger.info("Starting syncing mailchimp")
-        for configured_stream in catalog.streams:
-            yield from self._read_record(client=client, configured_stream=configured_stream, state=state)
-
-        logger.info("Finished syncing mailchimp")
-
-    def _client(self, config: json):
-        client = Client(username=config["username"], apikey=config["apikey"])
-
-        return client
-
-    @staticmethod
-    def _read_record(
-        client: Client, configured_stream: ConfiguredAirbyteStream, state: DefaultDict[str, any]
-    ) -> Generator[AirbyteMessage, None, None]:
-        entity_map = {"Lists": client.lists, "Campaigns": client.campaigns, "Email_activity": client.email_activity}
-
-        stream_name = configured_stream.stream.name
-
-        if configured_stream.sync_mode == SyncMode.full_refresh:
-            state.pop(stream_name, None)
-
-        for record in entity_map[stream_name](state=state):
-            yield record
+import base64
+import math
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import dateutil.parser
+import requests
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator
+from mailchimp3 import MailChimp
+
+
+class MailChimpStream(HttpStream, ABC):
+    url_base = "https://us2.api.mailchimp.com/3.0/"
+    primary_key = "id"
+    PAGINATION = 100
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # self.url_base = url_base #but `us2` can differ as {ds} in docs
+        self.current_offset = 0
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        api_data = decoded_response[self.data_field]
```

Comment:
`data_field` needs to be a @property,and an @abstract method, to ensure that child classes will define it.

also we could use `parse_response` here, as `list(self.parse_response(response))`